### DISCUSSION TITLE
Use "Configuring Rustfmt" with rust-analyzer on VS Code

### DISCUSSION THREAD

**Discussion by mirumirumi** [rust-lang/rust-analyzer#14601] [Q&A] [answered]
Hi!
I use rust-analyzer to fomat the code when saving (`editor.formatOnSave`).
 I want to customize the format to this format, so I figured I could apply the config here (https://rust-lang.github.io/rustfmt).

So I have tried the following settings in `settings.json` with various patterns, but it does not work fine.

```json
{
    "[rust]": {
        "editor.formatOnSave": true,
    },
      "rust-analyzer.rustfmt.extraArgs": [
        "blank_lines_upper_bound = 999",
        "imports_layout = \"HorizontalVertical\"",
        "imports_granularity = \"Crate\"",
        "overflow_delimited_expr = true",
        "group_imports = \"StdExternalCrate\"",
        "struct_lit_single_line = false",
    ],
}
```

And the trouble is, it seems that it only knows that I am trying to define an entry `rust-analyzer.rustfmt.extraArgs`, so the contents of the config are not reflected, and the code formatting function itself stops working.

What should I do?

**Comment by lnicola** [comment]
Can you use a [`rustfmt.toml`](https://rust-lang.github.io/rustfmt/)? The right syntax for `rust-analyzer.rustfmt.extraArgs` is probably `["--config", "blank_lines_upper_bound=999,overflow_delimited_expr=true"]`.

**Reply by mirumirumi** [reply, marked as answer]
Thanks for the response!

After much experimentation based on the advice I received, I found that I should not put spaces in the middle or escape `"`.
The following code is it works :)

```json
    "rust-analyzer.rustfmt.extraArgs": [
        "--config", "blank_lines_upper_bound=999,imports_layout=HorizontalVertical,imports_granularity=Crate,overflow_delimited_expr=true,group_imports=StdExternalCrate,struct_lit_single_line=false",
    ],
```

**Reply by lnicola** [reply]
I'd still suggest using a `rustfmt.toml` file instead.

**Reply by nulliver** [reply]
I know this is old but I also wanted to pass extra arguments to `rustfmt` and I wanted to use the `rustfmt.toml` file. I had some troubles making it work but what worked in the end was this:

I saved the rustfmt config file in `~/.config/rustfmt/rustfmt.toml`
```toml
tab_spaces = 2
```

In Zed's settings, I added this:
```json
"lsp": {
    "rust-analyzer": {
      "initialization_options": {
        "rustfmt": {
          "extraArgs": ["--config-path", "/Users/<your_user>/.config/rustfmt"]
        }
      }
    }
  }

```

Notice that in the configuration I had to use the absolute path to the `rustfmt.toml` file, I couldn't make it work with `$HOME` or `~` in the path to refer to my home directory.

**Reply by lnicola** [reply]
@nulliver according to https://rust-lang.github.io/rustfmt/ and https://docs.rs/dirs/latest/dirs/fn.config_dir.html, if you put it under `~/Library/Application Support`, you won't have to pass the path in the command line.

**Reply by nulliver** [reply]
@lnicola I tried that first and it didn't work but now that I got it working with the `extraArgs` configuration, I removed that configuration and moved the `rustftm.toml` file to `~/Library/Application Support/rustfmt/` and this time it worked! I guess I was doing something wrong the first time.

Thank you.
